### PR TITLE
Initial GI Bill Cmparison Tool backend integration

### PR DESCRIPTION
**Pull request by typesend**
- institution tables added *(may ultimately end up somewhere other than vets-api db)*
- migrated models, specs, and concerns
- created simple controller and routes
- added rake task and initial csv file
- autocomplete response corrected to not return null ids
- other tidying up

More PRs to come as I connect this up to the shiny new react frontend and improve tests.

See department-of-veterans-affairs/vets-website#3960 for necessary context.

### CODE REVIEW

File: app/models/concerns/tristateable.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+###############################################################################
+## Tristateable
+## Reads a column, bypassing activerecord field accessors, so that we can
+## differentiate between nil values and boolean false (by default nils in
+## activerecord are converted to false).
+##
+## Naturally, booleans can only hold true/false, so these fields should be
+## created as strings.
+###############################################################################
+module Tristateable
```
Comment (mphprogrammer) [inline app/models/concerns/tristateable.rb]: I think we should probably use a simpler comment style to be in line with the comment styles used in SM and RX apis. Its a nit-pick...

File: app/models/institution.rb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  }
+  DEGREES = {
+    0 => nil,
+    "ncd" => "Certificate",
+    1 => "Certificate",
+    "2-year" => 2,
+    2 => 2,
+    3 => 4,
+    4 => 4,
+    "4-year" => 4
```
Comment (mphprogrammer) [inline app/models/institution.rb]: Another nit-pick (sorry!) 
Double quoting will cause Rubocop to freak-out, the guidelines call for single quotes unless interpolating.

File: app/models/institution.rb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  # returns true if school is in USA
+  #
+  def in_usa?
+    country.try(:downcase) == 'usa'
+  end
+
+  # Gets the locale name correpsonding to the locale
+  #
+  def locale_name
+    LOCALE[locale] || "Locale Unknown"
```
Comment (mphprogrammer) [inline app/models/institution.rb]: Same Rubocop/double-quote nit-pick.

File: app/models/institution.rb [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  end
+
+  # Given a search term representing a partial school name, returns all
+  # schools starting with the search term.
+  #
+  def self.autocomplete(search_term)
+    search_term = search_term.to_s.strip.downcase
+    Institution.select("id, facility_code as value, institution as label")
+      .where("lower(institution) LIKE (?)", "#{search_term}%")
+  end
```
Comment (mphprogrammer) [inline app/models/institution.rb]: Same Rubocop/double-quote nit-pick

File: app/models/institution.rb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  ## schools might appear to be duplicated since they have the same name but
+  ## different facility codes.
+  #############################################################################
+  def self.search(search_term)
+    if search_term.empty?
+      @rset = Institution.with_type
+    else
+      search_term = search_term.to_s.downcase
+
+      clause = ["lower(facility_code) = (?) OR lower(institution) LIKE (?) OR lower(city) LIKE (?)"]
```
Comment (mphprogrammer) [inline app/models/institution.rb]: Same Rubocop/double-quote nit-pick

File: app/controllers/v0/institutions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    # e.g. /v0/gibct/institutions/autocomplete?term=harv
+    #
+    def autocomplete
+      search_term = params[:term]
+      render json: Institution.autocomplete(search_term)
+    end
+
+    # e.g. /v0/gibct/institutions/profile?facility_code=31800121
+    #
+    def profile
```
Comment (austin-martinez) [inline app/controllers/v0/institutions_controller.rb]: Could this be implemented as a `show` action with a route like `/v0/gibct/institutions/31800121`? Not sure how institutions will be accessed, but if it is only by facility code might make sense to map that to the identifier in routes

File: app/models/institution.rb [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+      rate = retention_all_students_otb.present? ? retention_all_students_otb : retention_all_students_ba
+    end
+
+    rate
+  end
+
+  # Returns the highest degree offered.
+  #
+  def highest_degree
+    DEGREES[pred_degree_awarded] || DEGREES[va_highest_degree_downcased] || "No Data"
```
Comment (mphprogrammer) [inline app/models/institution.rb]: Same Rubocop/double-quote nit-pick.

File: app/models/institution_type.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+class InstitutionType < ActiveRecord::Base
+  TYPES = {
+    "public" => "Public",
+    "ojt" => "On The Job Training",
+    "for profit" => "For Profit",
+    "private" => "Private",
+    "flight" => "Flight",
+    "correspondence" => "Correspondence",
+    "foreign" => "Foreign"
+  }
```
Comment (mphprogrammer) [inline app/models/institution_type.rb]: Same Rubocop/double-quote nit-pick.

File: lib/load_csv_helper.rb [hunk 9]
```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
```
Comment (mphprogrammer) [inline lib/load_csv_helper.rb]: Do you think we still need the load_csv_helper. I am not sure that this rake task would be terribly useful anymore, perhaps we should ignore it?

File: lib/tasks/gibct.rake [hunk 10]
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
```
Comment (mphprogrammer) [inline lib/tasks/gibct.rake]: Do you think we still need the rake. I am not sure that this rake task would be terribly useful anymore, perhaps we should ignore it?

File: lib/tasks/gibct.rake [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+require 'csv'
+require 'load_csv_helper'
+
+namespace :gibct do
+
+  desc "Clear and load institutions from CSV, e.g. rake gibct:load_csv[db/gibct_data.csv]"
```
Comment (mphprogrammer) [inline lib/tasks/gibct.rake]: Same Rubocop/double-quote nitpick.

File: lib/tasks/gibct.rake [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  task :load_csv, [:csv_file] => [:environment, :build_tables] do |t, args|
+    puts "Loading #{args[:csv_file]} ... "
+    count = 0
+
+    ActiveRecord::Base.transaction do
+      opts = {
+        headers: true,
+        encoding: "iso-8859-1:utf-8",
```
Comment (mphprogrammer) [inline lib/tasks/gibct.rake]: Same Rubocop/double-quote nitpick.

File: lib/tasks/gibct.rake [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  	puts "\n==== load_csv done!\n\n\n"
+  end
+
+  task build_tables: :environment do
+    if Rails.env.intern == :development
+      puts "Clearing development logs...\n"
+      Rake::Task['log:clear'].invoke
+    end
+
+    puts "Delete records in Institution in preparation for loading..."
```
Comment (mphprogrammer) [inline lib/tasks/gibct.rake]: Same Rubocop/double-quote nitpick.

File: lib/tasks/gibct.rake [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    Institution.delete_all
+
+    puts "Delete records in InstitutionType in preparation for loading..."
```
Comment (mphprogrammer) [inline lib/tasks/gibct.rake]: Same Rubocop/double-quote nitpick.

File: lib/tasks/gibct.rake [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    InstitutionType.delete_all
+
+    puts "Running migrations ..."
```
Comment (mphprogrammer) [inline lib/tasks/gibct.rake]: Same Rubocop/double-quote nitpick.

File: spec/factories/institutions.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+      country "usa"
+    end
+
+    trait :contains_harv do
+      sequence :institution do |n| ["hasharv#{n}", "has harv #{n}"].sample end
+      city "boston"
+      state "ma"
+      country "usa"
+    end
+  end
```
Comment (mphprogrammer) [inline spec/factories/institutions.rb]: Same Rubocop/double-quote nitpick.

File: spec/models/institution_spec.rb [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+  describe 'when evaluating a tristate boolean' do
+    let(:credit_for_mil_training) { build(:institution, credit_for_mil_training: nil) }
+    let(:vet_poc) { build(:institution, vet_poc: nil) }
+    let(:student_vet_grp_ipeds) { build(:institution, student_vet_grp_ipeds: nil) }
+    let(:soc_member) { build(:institution, soc_member: nil) }
+    let(:online_all) { build(:institution, online_all: nil) }
+    let(:caution_flag) { build(:institution, caution_flag: nil) }
+
+    it "nil is not false" do
```
Comment (mphprogrammer) [inline spec/models/institution_spec.rb]: Same Rubocop/double-quote nitpick.

File: spec/models/institution_spec.rb [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+
+    it "ojt institutions are not schools" do
+      inst.institution_type = ojt
+      exp = [inst.flight?, inst.correspondence?, inst.ojt?, inst.school?]
+      val = [false, false, true, false]
+
+      expect(exp).to eq(val)
+    end
+
+    it "other institutions are schools" do
```
Comment (mphprogrammer) [inline spec/models/institution_spec.rb]: Same Rubocop/double-quote nitpick.

**Comment by mphprogrammer** [comment]
Mostly-nit picky stuff about double quotes and Rubocop. The only point (other than the rake task and load csv not really being part of the vets-api world) is regarding the idea that we can maybe leverage the common collection models somehow - it may make your life easier on the front end. Just a thought...

**Comment by patrickvinograd** [comment]
You can just run rubocop -a and it will fix all the automatically-fixable things (tabs vs. spaces, quotes, etc).

File: app/controllers/v0/institutions_controller.rb (hunk 6 above)
Comment (mphprogrammer) [inline app/controllers/v0/institutions_controller.rb]: Yes the facility code is unique for every campus.

Comment (saneshark) [inline app/controllers/v0/institutions_controller.rb]: I will chime in and agree... in general my preference is to always stick to the standard controller actions and only introduce custom actions if it truly makes sense. This seems like a great candidate for a nested resource -- perhaps using `resource` vs `resources` if there is only 1 profile and not multiple.

**Comment by typesend** [comment]
Pushing code momentarily that addresses all the rubocop offenses and standardizes the routes.

Can we merge this PR? I'd like to implement additional units of work discussed in the next PRs.

**Comment by saneshark** [comment]
@typesend it looks like the test coverage below 90% for at least one file. You can test for these by running `bundle exec rake` locally. You can check the coverage on files by loading the html file that is automatically generated in the coverage folder after running rake.

**Comment by typesend** [comment]
Starting a new PR that incorporates this feedback and also changes related to how data is being fetched from the data app

**Pull request closed without merging**